src/braket/circuits/circuit.py

```diff
@@ -1217,16 +1229,73 @@ def _create_openqasm_header(
                 f"{serialization_properties.qubit_reference_type} supplied."
             )
 
-        frame_wf_declarations = self._generate_frame_wf_declarations()
+        frame_wf_declarations = self._generate_frame_wf_defcal_declarations(gate_calibrations)
         if frame_wf_declarations:
             ir_instructions.append(frame_wf_declarations)
         return ir_instructions
 
-    def _generate_frame_wf_declarations(self) -> Optional[str]:
-        frames = {}
-        waveforms = {}
+    def _validate_ngc_uniqueness(
+        self,
+        gate_calibrations: GateCalibrations,
+        frames: Dict[Frame],
+        waveforms: Dict[ArbitraryWaveform],
+    ) -> None:
+        for key, calibration in gate_calibrations.calibration_data.items():
+            for frame in calibration._frames.values():
+                _validate_uniqueness(frames, frame)
+                frames[frame.id] = frame
+            for waveform in calibration._waveforms.values():
+                _validate_uniqueness(waveforms, waveform)
+                waveforms[waveform.id] = waveform
+
+    def _generate_frame_wf_defcal_declarations(
+        self, gate_calibrations: Optional[GateCalibrations]
+    ) -> Optional[str]:
+        program = oqpy.Program(None)
+
+        frames, waveforms = self._get_frames_waveforms_from_instrs(gate_calibrations)
+
+        if gate_calibrations is not None:
+            self._validate_ngc_uniqueness(gate_calibrations, frames, waveforms)
+
+        # Declare the frames and waveforms across all pulse sequences
+        declarable_frames = [f for f in frames.values() if not f.is_predefined]
+        if declarable_frames or waveforms or gate_calibrations is not None:
+            frame_wf_to_declare = [f._to_oqpy_expression() for f in declarable_frames]
+            frame_wf_to_declare += [wf._to_oqpy_expression() for wf in waveforms.values()]
+            program.declare(frame_wf_to_declare, encal=True)
+
+            if gate_calibrations is not None:
+                for key, calibration in gate_calibrations.calibration_data.items():
+                    gate, qubits = key
+
+                    # ignoring parametric gates
+                    if hasattr(gate, "angle") and isinstance(gate.angle, FreeParameter):
+                        continue
```

Comment:
- Why are we ignoring parametric gates?
- Why do we need special handling for an angled gate? How does it extend to multi-angled gates in the future?